linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+}
+
+/** @brief `zero` divided by something is `zero` */
+template <typename T>
+MFEM_HOST_DEVICE constexpr zero operator/(zero, T /*other*/)
+{
+   return zero{};
+}
+
+/** @brief `zero` plus `zero` is `zero */
+MFEM_HOST_DEVICE constexpr zero operator+=(zero, zero) { return zero{}; }
+
+/** @brief `zero` minus `zero` is `zero */
+MFEM_HOST_DEVICE constexpr zero operator-=(zero, zero) { return zero{}; }
+
+/** @brief let `zero` be accessed like a tuple */
+template <int i>
+MFEM_HOST_DEVICE zero& get(zero& x)
+{
+   return x;
+}
+
+/** @brief the dot product of anything with `zero` is `zero` */
+template <typename T>
+MFEM_HOST_DEVICE zero dot(const T&, zero)
+{
+   return zero{};
+}
+
+/** @brief the dot product of anything with `zero` is `zero` */
+template <typename T>
+MFEM_HOST_DEVICE zero dot(zero, const T&)
+{
+   return zero{};
+}
+
+/**
+ * @brief Removes 1s from tensor dimensions
+ * For example, a tensor<T, 1, 10> is equivalent to a tensor<T, 10>
+ * @tparam T The scalar type of the tensor
+ * @tparam n1 The first dimension
+ * @tparam n2 The second dimension
+ */
+template <typename T, int n1, int n2 = 1>
+using reduced_tensor = typename std::conditional<
+                       (n1 == 1 && n2 == 1), double,
+                       typename std::conditional<n1 == 1, tensor<T, n2>,
+                       typename std::conditional<n2 == 1, tensor<T, n1>, tensor<T, n1, n2>
+                       >::type
+                       >::type
+                       >::type;
+
+/**
+ * @brief Creates a tensor of requested dimension by subsequent calls to a functor
+ * Can be thought of as analogous to @p std::transform in that the set of possible
+ * indices for dimensions @p n are transformed into the values of the tensor by @a f
+ * @tparam lambda_type The type of the functor
+ * @param[in] f The functor to generate the tensor values from
+ *
+ * @note the different cases of 0D, 1D, 2D, 3D, and 4D are implemented separately
+ *       to work around a limitation in nvcc involving __host__ __device__ lambdas with `auto` parameters.
+ */
+MFEM_SUPPRESS_NVCC_HOSTDEVICE_WARNING
+template <typename lambda_type>
+MFEM_HOST_DEVICE constexpr auto make_tensor(lambda_type f) ->
+tensor<decltype(f())>
+{
+   return {f()};
+}
+
+/**
+ * @brief Creates a tensor of requested dimension by subsequent calls to a functor
+ *
+ * @tparam n1 The dimension of the tensor
+ * @tparam lambda_type The type of the functor
+ * @param[in] f The functor to generate the tensor values from
+ * @pre @a f must accept @p n1 arguments of type @p int
+ *
+ * @note the different cases of 0D, 1D, 2D, 3D, and 4D are implemented separately
+ *       to work around a limitation in nvcc involving __host__ __device__ lambdas with `auto` parameters.
+ */
+MFEM_SUPPRESS_NVCC_HOSTDEVICE_WARNING
+template <int n1, typename lambda_type>
+MFEM_HOST_DEVICE auto make_tensor(lambda_type f) ->
+tensor<decltype(f(n1)), n1>
+{
+   using T = decltype(f(n1));
+   tensor<T, n1> A{};
```

Comment:
Value initialization is redundant.

Reply:
Initialization like this is a requirement to use this code in a `constexpr` context (which we had to remove because mfem still only supports C++11). I'd vote to leave these as-is, so it's easier to re-enable the `constexpr` feature when mfem eventually moves to a more modern C++ standard.